File: day03/solution.py
#!/usr/bin/env python3
import re

with open('input.txt', 'r') as f:
  data = f.read()

parsed = re.findall(r'mul\(([0-9]+),([0-9]+)\)', data)
# print(f'{parsed=}')
total = 0
for args in parsed:
  total += int(args[0]) * int(args[1])
print(f'part 1 {total=}')

parsed = re.findall(r'''(do\(\)|don't\(\)|mul\([0-9]+,[0-9]+\))''', data)
# print(f'{parsed=}')
enable = True
total = 0
for item in parsed:
  if item == 'do()': enable = True
  elif item == '''don't()''': enable = False
  elif enable:
    args = re.match(r'mul\(([0-9]+),([0-9]+)\)', item).groups()
    # print(f'{args=}')
    total += int(args[0]) * int(args[1])
print(f'part 2 {total=}')

File: day03/solution.ts
#!/usr/bin/env -S deno --allow-read

// const DEBUG = true;
const DEBUG = false;

const data = await Deno.readTextFile('input.txt');
if (DEBUG) console.debug(data);

const part1 = (data: string) => {
  const total = data.matchAll(/mul\(([0-9]+),([0-9]+)\)/g).reduce((acc, item) => {
    if (DEBUG) console.debug(item[1], item[2]);
    acc += parseInt(item[1]) * parseInt(item[2]);
    return acc;
  }, 0);
  console.log('part 1', total);
};

const part2 = (data: string) => {
  const total = data.matchAll(/(mul\(([0-9]+),([0-9]+)\)|do\(\)|don't\(\))/g).reduce(
    (acc, item) => {
      if (DEBUG) console.debug({ acc, item1: item[1] });
      switch (item[1]) {
        case 'do()':
          acc.enable = true;
          break;
        case "don't()":
          acc.enable = false;
          break;
        default:
          if (acc.enable) {
            acc.value += parseInt(item[2]) * parseInt(item[3]);
          }
      }
      return acc;
    },
    { value: 0, enable: true }
  );
  console.log('part 2', total.value);
};

part1(data);
part2(data);